File: reading_tracker/backend_api/app/schemas/book.py
from pydantic import BaseModel, Field


class CreateBookRequest(BaseModel):
    title: str = Field(min_length=1, max_length=255)
    author: str = Field(min_length=1, max_length=255)
    status: str = Field(min_length=1, max_length=50)
    rating: float | None = Field(default=None, ge=0, le=5)
    isbn: str | None = Field(default=None, max_length=32)
    pages: int | None = Field(default=None, ge=1, le=100000)
    publisher: str | None = Field(default=None, max_length=255)
    language_code: str | None = Field(default=None, min_length=2, max_length=10)
    cover_url: str | None = Field(default=None, max_length=2048)
    series_id: str | None = Field(default=None, max_length=255)
    volume: int | None = Field(default=None, ge=1, le=100000)
    genre_id: str | None = Field(default=None, max_length=255)
    age_category: str | None = Field(default=None, max_length=50)
    release_date: str | None = Field(default=None, max_length=10)
    format: str | None = Field(default=None, max_length=50)
    description: str | None = Field(default=None, max_length=10000)
    currentPage: int | None = Field(default=None, ge=1, le=100000)
    reading_start_date: str | None = Field(default=None, max_length=10)
    reading_end_date: str | None = Field(default=None, max_length=10)
    how_acquired: str | None = Field(default=None, max_length=100)
    where_acquired: str | None = Field(default=None, max_length=255)
    author_origin_id: str | None = Field(default=None, max_length=255)
    author_gender: str | None = Field(default=None, max_length=50)
    acquired_on: str | None = Field(default=None, max_length=10)
    price: float | None = Field(default=None, ge=0)
    notes: str | None = Field(default=None, max_length=10000)
    total_reading_minutes: int | None = Field(default=None, ge=0, le=10000000)
    first_publish_year: int | None = Field(default=None, ge=0, le=9999)


class UpdateBookRequest(BaseModel):
    title: str = Field(min_length=1, max_length=255)
    author: str = Field(min_length=1, max_length=255)
    status: str = Field(min_length=1, max_length=50)
    rating: float | None = Field(default=None, ge=0, le=5)
    isbn: str | None = Field(default=None, max_length=32)
    pages: int | None = Field(default=None, ge=1, le=100000)
    publisher: str | None = Field(default=None, max_length=255)
    language_code: str | None = Field(default=None, min_length=2, max_length=10)
    cover_url: str | None = Field(default=None, max_length=2048)
    series_id: str | None = Field(default=None, max_length=255)
    volume: int | None = Field(default=None, ge=1, le=100000)
    genre_id: str | None = Field(default=None, max_length=255)
    age_category: str | None = Field(default=None, max_length=50)
    release_date: str | None = Field(default=None, max_length=10)
    format: str | None = Field(default=None, max_length=50)
    description: str | None = Field(default=None, max_length=10000)
    currentPage: int | None = Field(default=None, ge=1, le=100000)
    reading_start_date: str | None = Field(default=None, max_length=10)
    reading_end_date: str | None = Field(default=None, max_length=10)
    how_acquired: str | None = Field(default=None, max_length=100)
    where_acquired: str | None = Field(default=None, max_length=255)
    author_origin_id: str | None = Field(default=None, max_length=255)
    author_gender: str | None = Field(default=None, max_length=50)
    acquired_on: str | None = Field(default=None, max_length=10)
    price: float | None = Field(default=None, ge=0)
    notes: str | None = Field(default=None, max_length=10000)
    total_reading_minutes: int | None = Field(default=None, ge=0, le=10000000)
    first_publish_year: int | None = Field(default=None, ge=0, le=9999)


class BookResponse(BaseModel):
    id: str
    user_id: str
    title: str
    author: str
    status: str
    rating: float | None
    isbn: str | None
    pages: int | None
    publisher: str | None
    language_code: str | None
    cover_url: str | None
    series_id: str | None
    volume: int | None
    genre_id: str | None
    age_category: str | None
    release_date: str | None
    format: str | None
    description: str | None
    currentPage: int | None
    reading_start_date: str | None
    reading_end_date: str | None
    how_acquired: str | None
    where_acquired: str | None
    author_origin_id: str | None
    author_gender: str | None
    acquired_on: str | None
    price: float | None
    notes: str | None
    total_reading_minutes: int | None
    first_publish_year: int | None
    created_at: str | None
    updated_at: str | None


class CreateReadingProgressRequest(BaseModel):
    page_number: int = Field(ge=0, le=100000)
    progress_date: str | None = Field(default=None, max_length=10)


class UpdateReadingProgressRequest(BaseModel):
    page_number: int = Field(ge=0, le=100000)
    progress_date: str = Field(min_length=10, max_length=10)


class ReadingProgressResponse(BaseModel):
    id: str
    book_id: str
    user_id: str
    progress_date: str
    page_number: int
    created_at: str | None
    updated_at: str | None

File: reading_tracker/lib/features/books/domain/models/book.dart
class Book {
  const Book({
    required this.id,
    required this.userId,
    required this.title,
    required this.author,
    required this.status,
    this.rating,
    this.isbn,
    this.pages,
    this.publisher,
    this.languageCode,
    this.coverUrl,
    this.seriesId,
    this.volume,
    this.genreId,
    this.ageCategory,
    this.releaseDate,
    this.format,
    this.description,
    this.currentPage,
    this.readingStartDate,
    this.readingEndDate,
    this.howAcquired,
    this.whereAcquired,
    this.authorOriginId,
    this.authorGender,
    this.acquiredOn,
    this.price,
    this.notes,
    this.totalReadingMinutes,
    this.firstPublishYear,
    this.createdAt,
    this.updatedAt,
  });

  final String id;
  final String userId;
  final String title;
  final String author;
  final String status;
  final double? rating;
  final String? isbn;
  final int? pages;
  final String? publisher;
  final String? languageCode;
  final String? coverUrl;
  final String? seriesId;
  final int? volume;
  final String? genreId;
  final String? ageCategory;
  final String? releaseDate;
  final String? format;
  final String? description;
  final int? currentPage;
  final String? readingStartDate;
  final String? readingEndDate;
  final String? howAcquired;
  final String? whereAcquired;
  final String? authorOriginId;
  final String? authorGender;
  final String? acquiredOn;
  final double? price;
  final String? notes;
  final int? totalReadingMinutes;
  final int? firstPublishYear;
  final String? createdAt;
  final String? updatedAt;

  factory Book.fromJson(Map<String, dynamic> json) {
    final ratingValue = json['rating'];
    final pagesValue = json['pages'];
    final volumeValue = json['volume'];
    final currentPageValue = json['currentPage'];
    final priceValue = json['price'];
    final totalReadingMinutesValue = json['total_reading_minutes'];
    final firstPublishYearValue = json['first_publish_year'];
    return Book(
      id: json['id'] as String,
      userId: json['user_id'] as String,
      title: json['title'] as String,
      author: json['author'] as String,
      status: json['status'] as String,
      rating: ratingValue is num ? ratingValue.toDouble() : null,
      isbn: json['isbn'] as String?,
      pages: pagesValue is num ? pagesValue.toInt() : null,
      publisher: json['publisher'] as String?,
      languageCode: json['language_code'] as String?,
      coverUrl: json['cover_url'] as String?,
      seriesId: json['series_id'] as String?,
      volume: volumeValue is num ? volumeValue.toInt() : null,
      genreId: json['genre_id'] as String?,
      ageCategory: json['age_category'] as String?,
      releaseDate: json['release_date'] as String?,
      format: json['format'] as String?,
      description: json['description'] as String?,
      currentPage: currentPageValue is num ? currentPageValue.toInt() : null,
      readingStartDate: json['reading_start_date'] as String?,
      readingEndDate: json['reading_end_date'] as String?,
      howAcquired: json['how_acquired'] as String?,
      whereAcquired: json['where_acquired'] as String?,
      authorOriginId: json['author_origin_id'] as String?,
      authorGender: json['author_gender'] as String?,
      acquiredOn: json['acquired_on'] as String?,
      price: priceValue is num ? priceValue.toDouble() : null,
      notes: json['notes'] as String?,
      totalReadingMinutes: totalReadingMinutesValue is num
          ? totalReadingMinutesValue.toInt()
          : null,
      firstPublishYear: firstPublishYearValue is num
          ? firstPublishYearValue.toInt()
          : null,
      createdAt: json['created_at'] as String?,
      updatedAt: json['updated_at'] as String?,
    );
  }
}

